### DISCUSSION TITLE
Context propagation not working with v1.7.0

### DISCUSSION THREAD

**Discussion by vimal3271** [open-telemetry/opentelemetry-java-instrumentation#4509] [Q&A] [answered]
I have a kafka producer and transformer app. I am using OTEL collector with Jaeger as backend. 

`JAVA_AGENT_OPTS=" $JAVA_AGENT_OPTS -Dotel.traces.exporter=otlp -Dotel.exporter.otlp.traces.protocol=grpc -Dotel.exporter.otlp.endpoint=http://otel-svc:4317"`

I am able to see the traces of producer and consumer. But they are not connected with same traceid. This was working fine till v1.5.3.

Any configuration that I need to enable.

**Comment by trask** [comment, marked as answer]
hi @vimal3271! the receive span now is the parent of the consumer span, and the receive span has a link to the producer span.

there's an option that may support the prior behavior, can you try with `-Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true`, and see if that works for you?

**Reply by vimal3271** [reply]
-Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true. This is working . Is this a bug in v17.0?

**Reply by mateuszrzeszutek** [reply]
> Is this a bug in v1.7.0?

No, this is in fact the correct behavior, according to OTel [messaging semantic conventions spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/messaging.md). Note that this semantic convention is not stable yet, and might change in the future.

**Reply by vimal3271** [reply]
Could you please elaborate ? I would like OTEL agent by default propagate the spans.

**Reply by trask** [reply]
Without `-Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true`, the agent should still be propagating spans, but instead of consumer.parent -> producer, it should be using links:

* consumer.parent -> receiver
* receiver.link -> producer

Unfortunately many backends do not support/render links yet though, which is why we also provide the option `-Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true`

**Reply by vimal3271** [reply]
Thank you @trask

**Reply by Asafb26** [reply]
for this case, what is the effect on the propagated baggage? will the consumer not going to receive the baggage sent by the producer?

**Reply by mateuszrzeszutek** [reply]
> for this case, what is the effect on the propagated baggage? will the consumer not going to receive the baggage sent by the producer?

That's an interesting case, we haven't considered that. No, by default the consumer will not get the baggage sent by the producer - just the span link is preserved. It should all work if you set the `-Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true` property though.

**Reply by wijnanjo** [reply]
So this new implementation (based on the current spec) makes it impossible (or hard) to correlate logs based on a common trace-id, right? In the case of a simple kafka producer and consumer, we get loglines with different trace-ids. 

Do we then need backends capable of stitching together the spans based on the parents _and_ links ? I might have it wrong but the new behavior seems to break logical flows.

**Comment by mvkrishna86** [comment]
Hi @trask , I am using Tempo/Grafana for storing and visualizing traces. I am having the same issue that the producer and consumer are having different traceids. 
I tried adding `-Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true`. But still I am getting the traceids differently. 
I am using java agent 1.6.0 jar file (auto instrumentation). Can you suggest any alternative please?

**Reply by trask** [reply]
hi @mvkrishna86! can you submit an issue with a repro that we can look at?

**Reply by mvkrishna86** [reply]
Thanks @trask , I am planning to migrate to 1.9.0. I will create the issue if I cant see the issue fixed.

**Reply by mvkrishna86** [reply]
Its working with 1.9.0. Thanks @trask .

**Reply by sukrit007wawa** [reply]
Tried 1.9.1 with  -Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true and do not see traceid getting propogated with ADOT Java  Auto instrumentation

From the trace,  I can confirm: `    "otel.resource.process.command_line": "/usr/lib/jvm/java-11-openjdk:bin:java -javaagent:/work/aws-opentelemetry-agent.jar -Djava.awt.headless=true -Djava.awt.headless=true -Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true"`

**Reply by rashid4lyf** [reply]
Same here, tried with 1.9.0 and the traceid is not getting propagated. I can confirm from the logs that the flag is applied at runtime: `java  -javaagent:/app/newrelic/newrelic.jar -javaagent:/app/opentelemetry-javaagent.jar -Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true -XX:+ExitOnOutOfMemoryError -Xms543m -Xmx543m -XX:MetaspaceSize=145m -XX:MaxMetaspaceSize=145m -jar /app/wh-kafka-standard.jar`

**Reply by trask** [reply]
@sukrit007wawa @rashid4lyf can you open a github issue and provide a repro of the problem(s) you are seeing?

**Comment by PhoenixFlash10** [comment]
Do you have any documentation around this - **-Dotel.instrumentation.common.experimental.suppress-messaging-receive-spans=true**?
I couldn't find anything on https://opentelemetry.io/docs/instrumentation/java/automatic/agent-config/

@trask @vimal3271

**Reply by laurit** [reply]
@PhoenixFlash10 it is documented in https://opentelemetry.io/docs/instrumentation/java/automatic/agent-config/#capturing-consumer-message-receive-telemetry-in-messaging-instrumentations

**Reply by PhoenixFlash10** [reply]
thanks for your reply @laurit 
but I don't see that specific property, the one you pointed to is - otel.instrumentation.messaging.experimental.receive-telemetry.enabled
the one mentioned in this answer is - otel.instrumentation.common.experimental.suppress-messaging-receive-spans

are you trying to say that they are the same?

**Reply by PhoenixFlash10** [reply]
The thing is I am facing the same issue. let me try to mention my use-case here :

Kafka-Topic-1 (kt1) ->
Sub-Component-1 (sc1) [listens-processes-publishes] ->
Kafka-Topic-2 (kt2) ->
Sub-Component-2 (sc2) [listens-processes-publishes] ->
Kafka-Topic-3 (kt3) ->
Sub-Component-3 (sc3) [listens-processes-persistsToMongo]

**Note(s)/Issue(s) :**
1. (sc2) consolidates multiple messages from (kt2) and processes them together
2. We are trying to use OpenTelemetry Java agent
3. Also, we are manipulating other Kafka message headers[mentioning this, to understand a possibility of traces getting lost if we manipulate headers]
4. When I try to visualize this up on Grafana with Jaeger as my datasource, I see the traceIDs are not being related!

Can you please help? @laurit @trask @vimal3271

**Reply by PhoenixFlash10** [reply]
Created a stack question as well - https://stackoverflow.com/questions/77711724/contexttraceid-and-spanid-getting-lost-when-using-kafka-for-micro-processing

**Reply by laurit** [reply]
> but I don't see that specific property, the one you pointed to is - otel.instrumentation.messaging.experimental.receive-telemetry.enabled the one mentioned in this answer is - otel.instrumentation.common.experimental.suppress-messaging-receive-spans
> 
> are you trying to say that they are the same?

@PhoenixFlash10 perhaps you didn't notice but this discussion is more than 2 years old. The property referenced here was renamed a long time ago. Also note that the current default behavior is as if `suppress-messaging-receive-spans` was set to `true`. If you wish for somebody to look at this your best bet is to open an new issue and include a minimal sample app that reproduces your problem.

**Reply by PhoenixFlash10** [reply]
@laurit thanks for your reply again
Does that mean if I set "otel.instrumentation.messaging.experimental.receive-telemetry.enabled" to **true**, it will copy the traceID and spanID from the producer?
or, I should set it to **false** and then will the traceID and spanID be copied?

**Reply by laurit** [reply]
> Does that mean if I set "otel.instrumentation.messaging.experimental.receive-telemetry.enabled" to true, it will copy the traceID and spanID from the producer?

No it does not. It means that producer and consumer will be in separate traces that are connected with a span link. The default (`receive-telemetry.enabled=false`) has producer and consumer in the same trace.

**Reply by PhoenixFlash10** [reply]
can I get a few more details around how the instrumentation for Kafka messages work?
The reason for this question is - if there are multiple messages getting consolidated and read, how is the log tracing through Jaeger going to work?

**Reply by laurit** [reply]
> can I get a few more details around how the instrumentation for Kafka messages work?

You can examine the source code of the instrumentation at https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/kafka/kafka-clients

> The reason for this question is - if there are multiple messages getting consolidated and read, how is the log tracing through Jaeger going to work?

I guess it depends on what you mean by consolidating. Kafka instrumentation works best when you read a `ConsumerRecord`, act on it (i.e. do something that produces spans) and proceed to the next message.  For example if you read a `ConsumerRecord`, put it into a list, do the same with other messages and finally process the list in a way that produces spans then it most likely won't work unless you manually restore the context for each `ConsumerRecord`.

**Reply by PhoenixFlash10** [reply]
ohh ok ok. This makes sense to me now. Let me check few more details around this and get back to you!
